perf(minifier/minimize_statements): reduce allocations of `Vec`

Dunqing opened the pull request:
Use `ctx.ast.vec_from_iter` to allocate a `Vec` for the result of `drain` and use `take_in` to take a `Vec` rather than use `vec_from_iter` + `drain` to avoid intermediate allocation.

<img width="742" alt="image" src="https://github.com/user-attachments/assets/10011073-6706-435a-8d79-0b805e14f20c" />

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10435?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10435** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10435?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10435?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10434** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10434?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10418** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10418?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#9881](https://github.com/oxc-project/oxc/pull/9881) <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9881?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10435).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Actually I think this might be even better:

Before:

```rs
let drained_stmts = stmts.drain(i + 1..);
let mut body = if let Some(alternate) = if_stmt.alternate.take() {
    ctx.ast.vec_from_iter(iter::once(alternate).chain(drained_stmts))
} else {
    ctx.ast.vec_from_iter(drained_stmts)
};
```

After:

```rs
let mut body = if let Some(alternate) = if_stmt.alternate.take() {
    let drained_stmts = stmts.drain(i + 1..);
    ctx.ast.vec_from_iter(iter::once(alternate).chain(drained_stmts))
} else {
    stmts.split_off(i + 1)
};
```

`split_off` avoids using `vec_from_iter`. It knows the exact required capacity of the new `Vec`.

Similar to my comment on preceding PR https://github.com/oxc-project/oxc/pull/10434#issuecomment-2809350194, later on we can make our `split_off` implementation more efficient, not copying any data or performing any allocations.

Then this simpler version would be equally efficient (or possibly more, as if `stmts` has spare capacity, it avoids an allocation):

```rs
let mut body = stmts.split_off(i + 1);
if let Some(alternate) = if_stmt.alternate.take() {
    body.insert(0, alternate);
}
```

Dunqing committed:
perf(minifier/minimize_statements): reduce allocations of `Vec` (#10435)

Use `ctx.ast.vec_from_iter` to allocate a `Vec` for the result of `drain` and use `take_in` to take a `Vec` rather than use `vec_from_iter` + `drain` to avoid intermediate allocation.

<img width="742" alt="image" src="https://github.com/user-attachments/assets/10011073-6706-435a-8d79-0b805e14f20c" />
